src/ynEIGEN/RedemptionAssetsVault.sol

```diff
@@ -0,0 +1,230 @@
+                }
+            }
+        }
+
+        emit AssetTransferred(ETH_ASSET, msg.sender, to, amountTransferred);
+    }
+
+    /** 
+     * @notice Withdraws a specified amount of redemption assets and processes them through ynETH.
+     * @param amount The amount of ETH to withdraw and process.
+     * @dev Requires the caller to be the redeemer and the contract to not be paused.
+     */
+    function withdrawRedemptionAssets(uint256 amount) public onlyRedeemer whenNotPaused nonReentrant {
+        IERC20[] memory assets = assetRegistry.getAssets();
+        for (uint256 i = 0; i < assets.length; ++i) {
+            IERC20 asset = assets[i];
+            uint256 assetBalance = balances[address(asset)];
+            if (assetBalance > 0) {
+                uint256 assetBalanceInUnit = assetRegistry.convertToUnitOfAccount(asset, assetBalance);
+                if (assetBalanceInUnit >= amount) {
+                    uint256 reqAmountInAsset = assetRegistry.convertFromUnitOfAccount(asset, amount);
+                    ynEigen.processWithdrawn(reqAmountInAsset, address(asset));
+                    balances[address(asset)] -= reqAmountInAsset;
+                    break;
+                } else {
+                    ynEigen.processWithdrawn(assetBalance, address(asset));
+                    balances[address(asset)] = 0;
+                    amount -= assetBalanceInUnit;
+                }
+            }
+        }
+        emit AssetWithdrawn(ETH_ASSET, msg.sender, address(ynEigen), amount);
+    }
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  MODIFIERS  ---------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    /** 
+     * @notice Ensure that the given address is not the zero address.
+     * @param _address The address to check.
+     */
+    modifier notZeroAddress(address _address) {
+        if (_address == address(0)) {
+            revert ZeroAddress();
+        }
+        _;
+    }
+
+    /** 
+     * @notice Checks if the contract is not paused.
+     */
+    modifier whenNotPaused() {
+        if (paused) {
+            revert ContractPaused();
+        }
+        _;
+    }
+
+    /**
+     * @notice Ensures that the caller has the REDEEMER_ROLE.
+     */
+    modifier onlyRedeemer() {
+        if (msg.sender != redeemer) {
+            revert NotRedeemer(msg.sender);
+        }
+        _;
+    }
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  PAUSE FUNCTIONS  ---------------------------------
+    //--------------------------------------------------------------------------------------
+
+    /** 
+     * @notice Pauses the contract, preventing certain actions.
+     */
+    function pause() external onlyRole(PAUSER_ROLE) {
+        paused = true;
```

Comment:
event?

Reply:
and what about `require(!paused)`?

Reply:
agreed

Reply:
https://github.com/yieldnest/yieldnest-protocol/pull/164/commits/cb91223d259e2472451a0d27868aef4cc28c0bd2